modules/core/src/main/java/org/apache/ignite/internal/processors/odbc/SqlListenerUtils.java

```diff
@@ -246,12 +274,61 @@ else if (cls == Timestamp[].class)
             writer.writeTimestampArray((Timestamp[])obj);
         else if (cls == java.util.Date[].class || cls == java.sql.Date[].class)
             writer.writeDateArray((java.util.Date[])obj);
+        else if (obj instanceof SqlInputStreamWrapper)
+            writeInputStreamAsByteArray(writer, (SqlInputStreamWrapper)obj);
+        else if (obj instanceof Blob)
+            writeBlobAsByteArray(writer, (Blob)obj);
         else if (binObjAllow)
             writer.writeObjectDetached(obj);
         else
             throw new BinaryObjectException("Custom objects are not supported");
     }
 
+    /**
+     * Write byte array from the InputStream enclosed in the stream wrapper.
+     *
+     * @param writer Writer.
+     * @param wrapper stream wrapper
+     */
+    private static void writeInputStreamAsByteArray(BinaryWriterExImpl writer, SqlInputStreamWrapper wrapper)
+            throws BinaryObjectException {
+        InputStream in = wrapper.getInputStream();
+        Integer len = wrapper.getLength();
+
+        if (len == null)
+            writer.writeByteArrayFromInputStream(in);
+        else {
+            int written = writer.writeByteArrayFromInputStream(in, len);
+
+            if (len != written)
+                throw new BinaryObjectException("Input stream length mismatch. [declaredLength=" + len + ", " +
+                        "actualLength=" + written + "]");
+        }
+    }
+
+    /**
+     * Write byte array from the Blob instance.
+     *
+     * @param writer Writer.
+     * @param blob Blob.
+     */
+    private static void writeBlobAsByteArray(BinaryWriterExImpl writer, Blob blob)
+            throws BinaryObjectException {
+        try {
+            int len = (int)blob.length();
+            InputStream in = blob.getBinaryStream(1, len);
+
+            int written = writer.writeByteArrayFromInputStream(in, len);
+
+            if (len != written)
```

Comment:
parentess for multiline `if`.

if (len != written) {
throw new BinaryObjectException("Blob length mismatch. [declaredLength=" + len + ", " +
                        "actualLength=" + written + "]");
}